cli/src/test.rs

```diff
@@ -32,166 +24,48 @@ pub struct TestArgs {
-        .as_ref()
-        .map(|r| r.exit_code)
-        .unwrap_or(run_exit_code);
-
-    let exec_start = run_result.exec_start;
-    if let Err(e) = run_upload(
-        upload_args,
-        Some(command.join(" ")),
-        None, // don't re-run quarantine checks
-        codeowners,
-        exec_start,
-    )
-    .await
-    {
-        log::error!("Error uploading test results: {:?}", e)
-    };
-
-    Ok(exit_code)
-}
-
-pub enum JunitSpec {
-    Paths(Vec<String>),
-    BazelBep(String),
+#[derive(Debug, Clone)]
+pub struct TestRunResult {
+    pub command: String,
+    pub exec_start: SystemTime,
+    pub exit_code: i32,
 }
 
-pub async fn run_test_command(
-    repo: &BundleRepo,
-    command: &String,
-    args: Vec<&String>,
-    junit_spec: JunitSpec,
-    team: Option<String>,
-    codeowners: &Option<CodeOwners>,
-) -> anyhow::Result<RunResult> {
-    let start = SystemTime::now();
-    let exit_code = run_test_and_get_exit_code(command, args)?;
-    log::info!("Command exit code: {}", exit_code);
+pub async fn run_test(
+    TestArgs {
+        upload_args,
+        command,
+    }: TestArgs,
+) -> anyhow::Result<i32> {
+    let pre_test_context = gather_pre_test_context(upload_args.clone())?;
 
-    let output_paths = match junit_spec {
-        JunitSpec::Paths(paths) => paths
-            .into_iter()
-            .map(JunitReportFileWithStatus::from)
-            .collect(),
-        JunitSpec::BazelBep(bep_path) => {
-            let mut parser = BazelBepParser::new(bep_path);
-            let bep_result = parser.parse()?;
-            print_bep_results(&bep_result);
-            bep_result.uncached_xml_files()
-        }
-    };
+    log::info!("running command: {:?}", command);
+    let test_run_result = run_test_command(&command).await?;
+    let test_run_result_exit_code = test_run_result.exit_code;
 
-    let file_set_builder = FileSetBuilder::build_file_sets(
-        &repo.repo_root,
-        &output_paths,
-        team,
-        codeowners,
-        Some(start),
-    )?;
+    let upload_run_result =
+        run_upload(upload_args, Some(pre_test_context), Some(test_run_result)).await;
 
-    Ok(RunResult {
-        exit_code,
-        file_set_builder,
-        exec_start: Some(start),
+    upload_run_result.or_else(|e| {
+        log::error!("Error uploading test results: {:?}", e);
+        Ok(test_run_result_exit_code)
     })
 }
 
-fn run_test_and_get_exit_code(command: &String, args: Vec<&String>) -> anyhow::Result<i32> {
-    let mut child = Command::new(command)
-        .args(args)
+pub async fn run_test_command<T: AsRef<str>>(command: &[T]) -> anyhow::Result<TestRunResult> {
+    let exec_start = SystemTime::now();
+    let mut child = Command::new(command.first().map(|s| s.as_ref()).unwrap_or_default())
+        .args(
+            command
+                .iter()
+                .skip(1)
+                .map(|s| s.as_ref())
+                .collect::<Vec<_>>(),
+        )
```

Comment:
What was the reason for this change and the generic?

Reply:
`&String` is an odd type since that's usually represented by `&str`, but to make whatever calls this even more simplified we can accept anything that can be referenced as `&str` with the generic `AsRef<str>`. It's essentially a bit of clean up.

This is very opinionated, so I'm open to doing things differently. Here's a bit more background:
https://steveklabnik.com/writing/when-should-i-use-string-vs-str/